### PR TITLE
[WIP] Giant refactor!!!!!!!!!

### PR DESCRIPTION
**Pull request by apg**
(Yes, this is like... the 10th giant refactor of this all, and has not been tested against an actual app.)

Remove any and all assumptions that things are request driven. It's
all periodic, all the time.

In addition, simplify the concept of a meter, into a panel which
contains instruments that can be `instrument!`'d. These `!`methods
modify passed in state and are essentially folded over to collect
the current values of the gauges and counters. The instruments
themselves emit the proper key, and in the case of counters that
we'd like to treat as special counters, they apply sampling. See
an explanation of that in doc.rb.

Fix up the Railtie such that we can actually configure this thing
a bit easier, and remove things we'll just plain not utilize, like
the ChangeInstrument and GaugeInstrument which weren't utilized
and confused things.

### CODE REVIEW

**Comment by amerine** [comment]
Neat.

**Comment by apg** [comment]
Connects to #3

File: lib/barnes/railtie.rb [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
+      :panels => [],
+    }
 
     initializer 'barnes' do |app|
       require 'statsd'
 
-      app.config.barnes.logger ||= Rails.logger
-      Periodic.new config.barnes
-    end
+      sample_rate = config.barnes[:interval] / 60.0
```
Comment (amerine) [inline lib/barnes/railtie.rb]: Why 60.0?

File: lib/barnes/reporter.rb [hunk 2]
```diff
@@ -22,23 +14,14 @@ def report(env)
     def report_statsd(env)
       method = @statsd.respond_to?(:easy) ? :easy : :batch
       @statsd.send(method) do |statsd|
-        send_to_statsd statsd, :count, @counter_sample_rate, env[Barnes::COUNTERS], :'Rack.Server', @counter_dimensions.call(env)
-        send_to_statsd statsd, :gauge, @gauge_sample_rate, env[Barnes::GAUGES], :'Rack.Server', @gauge_dimensions.call(env)
+        send_to_statsd statsd, :count, @sample_rate, env[Barnes::COUNTERS], :'Rack.Server.All'
+        send_to_statsd statsd, :gauge, 1.0, env[Barnes::GAUGES], :'Rack.Server.All'
```
Comment (amerine) [inline lib/barnes/reporter.rb]: Why 1.0?

**Review by amerine — APPROVED** [review]
This looks excellent to me, I think the switch to purely periodic is nice. I don't know enough about the reasoning behind this refactor to comment on the change, but I do wonder if there is something about the properties of tracking out-of-band (e.g. non-request driven) GC that we lose by using a purely periodic system?

I'm a bit curious about some hard-coded values, but beyond that 👍.

File: lib/barnes/railtie.rb (hunk 1 above)
Comment (apg) [reply lib/barnes/railtie.rb]: It should probably not be hardcoded, I agree. We're using a 60s aggregation period. This should certainly be configurable with this as default.

File: lib/barnes/reporter.rb (hunk 2 above)
Comment (apg) [reply lib/barnes/reporter.rb]: Sample rates for gauges in statsd have no meaning, but the method requires us to provide it. See: https://github.com/etsy/statsd/blob/master/stats.js#L269. It's probably a good thing to document.

**Comment by apg** [comment]
@amerine one thing that trashed does that's sort of strange is assign an arbitrary sample rate to the per request data. So, I'm not convinced it's very accurate, and in order to support it, there's quite a lot of complexity in the code that as an outsider (Ruby is by far not my first language) I had trouble piecing together.

Heroku metrics doesn't currently have any understanding of per request metrics (we aggregate 5s of requests into a histogram for response times before they even hit Kafka) and so per request metrics just don't currently *fit* in our model. As we get more sophisticated, I think there may be some opportunity for us to do sampling of request metrics that align with full HTTP requests. One could imagine that we instrument all requests whose request id ends in '0', or something silly like that in order to provide much greater visibility.

**Review by schneems — APPROVED** [review]

**Pull request merged**